### DISCUSSION TITLE
Flexcache Status

### DISCUSSION THREAD

**Discussion by jmg011** [NetApp/harvest#979] [Q&A] [answered]
Hi,

I have enabled flexcache metrics, but I do not have a way to get the flexcache status (connected, connecting...etc.)

These APIs do not give status information.

-  flexcache-connected-cache-get-iter
-  flexcache-get-iter
-  flexcache-origin-config-get-iter

I can get the status from volume metric, but then it is only set for online or not online. How do I get more breakdown on flexcache status?

**Comment by cgrinds** [comment, marked as answer]
hi @jmg011 we'll take a look. I think you're right that this information is returned via `volume-get-iter` response `volume-state/state` but I want to check what the REST endpoint is doing because it includes `origins.state` usable like this `https://10.231.84.110/api/storage/flexcache/flexcaches?fields=origins.state`

**Reply by jmg011** [reply]
@cgrinds Let me know if you find this in zapi.

**Reply by cgrinds** [reply]
@jmg011 the only thing I've found is `volume-get-iter` and the response `netapp/results/attributes-list/volume-attributes/volume-state-attributes/state` which you already mentioned.

@ruanruijuan also thinks that the `volume-state-attributes` are applicable to ALL volumes, while `origin.state` in rest is only for the flexcache ones, which means it isn't a true one to one mapping.

**It's probably better to gather this from REST instead of ZAPI**

Example shown below
 
## REST
```sh
curl -sk --netrc "https://10.231.84.110/api/storage/flexcache/flexcaches?fields=origins.state&name=sri_vol_flexcache3"
{
  "records": [
    {
      "uuid": "e88ecbb5-4ec1-11eb-8371-00a098e0060e",
      "name": "sri_vol_flexcache3",
      "origins": [
        {
          "state": "online"
        }
      ],
    }
  ],
  "num_records": 1
}                                            
```

## ZAPI that likely is not one to one with the REST call
``` 
<netapp xmlns="http://www.netapp.com/filer/admin" version="1.160">
  <volume-get-iter>
    <query>
      <volume-attributes>
        <volume-id-attributes>
          <name>sri_vol_flexcache3</name>
        </volume-id-attributes>
      </volume-attributes>
    </query>
    <desired-attributes>
      <volume-attributes>
        <volume-id-attributes>
          <name/>
        </volume-id-attributes>     
        <volume-state-attributes>
          <state/>
        </volume-state-attributes>
      </volume-attributes>
    </desired-attributes>
  </volume-get-iter>
</netapp>
```

which returns

```
<netapp xmlns="http://www.netapp.com/filer/admin" version="1.211">
  <results status="passed">
    <attributes-list>
      <volume-attributes>
        <volume-id-attributes>
          <name>sri_vol_flexcache3</name>
          <owning-vserver-name>vs8_mixed</owning-vserver-name>
        </volume-id-attributes>
        <volume-state-attributes>
          <state>online</state>                       <=============== matches origins.state
        </volume-state-attributes>
      </volume-attributes>
    </attributes-list>
    <num-records>1</num-records>
  </results>
</netapp>
```

**Reply by jmg011** [reply]
@cgrinds Thank you. Can we have collection through both ZAPI and REST? Or it has to be one or the other?

**Reply by rahulguptajss** [reply]
@jmg011 Harvest does not prevent you from collecting the same resource with ZAPI and REST. Though we recommend to enable only one of these for a relevant resource. This will help avoid duplicate records in dashboard.

**Reply by jmg011** [reply]
@rahulguptajss Thank you. Yes, the idea is to collect the objects with REST if those are not available through ZAPI. There won't be duplicates if I precisely build the metrics which are unique. I will test it.

**Reply by ruanruijuan** [reply]
The volume state from the ZAPI provides the state for all volumes, while the REST provides the origins.state for the flexcahe ones. So the REST is better here.
